backend/src/layer/logic/image_review/image_reviewer.rs

```
@@ -0,0 +1,239 @@
+
+pub struct ImageReviewer<D, H>
+where
+    D: ImageReviewDataAccess,
+    H: ImageHoster,
+{
+    data_access: D,
+    image_hoster: H,
+}
+
+#[async_trait]
+impl<D, H> ImageReviewScheduling for ImageReviewer<D, H>
+where
+    D: ImageReviewDataAccess,
+    H: ImageHoster,
+{
+    async fn start_image_review(&self) {
+        let today = Local::now().date_naive();
+        self.review_images(
+            self.data_access
+                .get_n_images_by_rank_date(NUMBER_OF_IMAGES_TO_CHECK, today)
+                .await,
+        )
+        .await;
+        self.review_images(
+            self.data_access
+                .get_n_images_next_week_by_rank_not_checked_last_week(NUMBER_OF_IMAGES_TO_CHECK)
+                .await,
+        )
+        .await;
+        self.review_images(
+            self.data_access
+                .get_n_images_by_last_checked_not_checked_last_week(NUMBER_OF_IMAGES_TO_CHECK)
+                .await,
+        )
+        .await;
+    }
+}
+
+impl<D, H> ImageReviewer<D, H>
+where
+    D: ImageReviewDataAccess,
+    H: ImageHoster,
+{
+    #[must_use]
+    pub const fn new(data_access: D, image_hoster: H) -> Self {
+        Self {
+            data_access,
+            image_hoster,
+        }
+    }
+
+    async fn review_images(&self, images: Result<Vec<Image>>) {
+        if let Err(error) = self.try_review_images(images).await {
+            warn!("{error:#?}");
+        }
+    }
+
+    async fn try_review_images(&self, images: Result<Vec<Image>>) -> ReviewerResult<()> {
+        let images = images.map_err(|_e| ReviewerError::ImageGetError)?;
+        for image in images {
+            self.review_image(&image).await?;
+        }
+        Ok(())
+    }
+
+    async fn review_image(&self, image: &Image) -> ReviewerResult<()> {
+        let exists = self
+            .image_hoster
+            .check_existence(&image.image_hoster_id)
+            .await
+            .map_err(|_e| ReviewerError::CheckError(image.id))?;
```

we dont want to discard the error but inject it in to our ReviewerError. This error is probably more interesting than the image id so you may just make the `CheckError` be created ftom the error of the image hoster using `From`.